src/reader_disc.cpp

```diff
@@ -0,0 +1,559 @@
+      break;
+    }
+  } else {
+    printf("Unsupported media!\n");
+    goto exit_reader;
+  }
+
+  if (parm->pass >= PASS_RECOVER)
+    scnt = 1;
+
+  map->lock();
+  if (parm->pass < PASS_RECOVER) {
+    map->get_next(&lba, BM_WAIT, &scnt);
+  } else {
+    bool rce, wce;
+    printf("Recover pass - trying to disable read cache...\n");
+    get_cache(dev, &rce, &wce);
+    if (!rce) {
+      printf("Read cache already disabled:)\n");
+    } else {
+      if (set_cache(dev, 0, wce)) {
+        printf("It seems drive does not support disabling read cache!\n");
+      } else {
+        printf("Read cache disabled successfully\n");
+      }
+    }
+    map->get_next(&lba, BM_FAIL, &scnt);
+  }
+  map->set(lba, BM_READ, scnt);
+  map->unlock();
+
+  if (dev->media.dvdcss.protection == 0x01) {
+    for (int i = 0; i < DVD_KEY_SIZE; i++)
+      dev->media.dvdcss.TK[i] = 0;
+    p_title = dev->media.dvdcss.p_titles;
+    descramble = 0;
+    if (!p_title) {
+      title_next = 0xFFFFFFFF;
+    } else {
+      title_next = p_title->i_startlb;
+    }
+    printf("\nSetting title key: %02X:%02X:%02X:%02X:%02X  for %x-%x, CSS=%d\n",
+           dev->media.dvdcss.TK[0], dev->media.dvdcss.TK[1],
+           dev->media.dvdcss.TK[2], dev->media.dvdcss.TK[3],
+           dev->media.dvdcss.TK[4], 0, title_next, descramble);
+  }
+
+  gettimeofday(&t_e, NULL);
+  seek(dev, 0);
+  while (lba < dev->media.capacity) {
+    if (lba == 0xFFFFFFFF)
+      goto exit_reader;
+    t_s.tv_sec = t_e.tv_sec;
+    t_s.tv_usec = t_e.tv_usec;
+
+    if (dev->media.type & DISC_CD) {
+      read(dev, dev->rd_buf, lba, scnt);
+    } else if (dev->media.type & DISC_DVD) {
+      if (dev->media.dvdcss.protection == 0x01) {
+
+        if (lba >= title_next && p_title) {
+          //	scnt = 16;
+          while (lba >= title_next) {
+            if (p_title->p_next) {
+              title_next = p_title->p_next->i_startlb;
+            } else {
+              title_next = 0xFFFFFFFF;
+            }
+            if (lba >= title_next)
+              p_title = p_title->p_next;
+          }
+
+          memcpy(dev->media.dvdcss.TK, p_title->p_key, sizeof(dvd_key_t));
+          if (!memcmp(dev->media.dvdcss.TK, "\0\0\0\0\0", 5))
+            descramble = 0;
+          else
+            descramble = 1;
+          printf("\nSetting title key: %02X:%02X:%02X:%02X:%02X  for %x-%x, "
+                 "CSS=%d\n",
+                 dev->media.dvdcss.TK[0], dev->media.dvdcss.TK[1],
+                 dev->media.dvdcss.TK[2], dev->media.dvdcss.TK[3],
+                 dev->media.dvdcss.TK[4], lba, title_next, descramble);
+          p_title = p_title->p_next;
+        }
+
+        read_dvd(dev, dev->rd_buf, lba, scnt,
+                 descramble ? DVDCSS_READ_DECRYPT : 0);
+      } else {
+        read(dev, dev->rd_buf, lba, scnt);
+      }
+    }
+    gettimeofday(&t_e, NULL);
+
+    printf("%s: %5.2f X, lba %7d / %7d  ( %06x / %06x ) scnt=%d\r", dev->device,
```

Comment:
Codacy found an issue: [%d in format string (no. 3) requires 'int' but the argument type is 'unsigned int'.](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)